src/main/java/org/openjdk/shenandoah/ShenandoahVisualizer.java

```diff
@@ -22,79 +23,354 @@
+                    return;
+                }
+            } else if (arg.equals("-logFile")) {
+                if (i < args.length) {
+                    isReplay = true;
+                    filePath[0] = args[i++];
+                } else {
+                    System.out.println("-logFile requires a file path");
+                    return;
+                }
+            } else {
+                System.out.println("ShenandoahVisualizer: Illegal option " + arg);
+                System.out.println("Usage: [-vm vmIdentifier] [-logFile filePath]");
+                return;
+            }
         }
+        //
 
         JFrame frame = new JFrame();
         frame.setLayout(new GridBagLayout());
         frame.setTitle("Shenandoah GC Visualizer");
         frame.setSize(INITIAL_WIDTH, INITIAL_HEIGHT);
 
-        DataProvider data = new DataProvider(args[0]);
+        final RenderRunner renderRunner;
+        ToolbarPanel toolbarPanel = new ToolbarPanel(isReplay);
+        int totalSnapshotSize = 0;
+
+        if (isReplay) {
+            DataLogProvider data = new DataLogProvider(filePath[0]);
+            totalSnapshotSize = data.getSnapshotsSize();
+            toolbarPanel.setSize(totalSnapshotSize);
+            toolbarPanel.setSnapshots(data.getSnapshots());
+            renderRunner = new RenderRunner(data, frame, toolbarPanel);
+            toolbarPanel.setModeField(PLAYBACK);
+            toolbarPanel.setEnabledRealtimeModeButton(true);
+            toolbarPanel.setFileNameField(filePath[0]);
+        } else {
+            DataProvider data = new DataProvider(vmIdentifier);
+            renderRunner = new RenderRunner(data, frame, toolbarPanel);
+            toolbarPanel.setModeField(REALTIME);
+            toolbarPanel.setEnabledRealtimeModeButton(false);
+        }
+
+        // Executors
+        ScheduledExecutorService service = Executors.newScheduledThreadPool(1);
+        final ScheduledFuture<?>[] f = {service.scheduleAtFixedRate(renderRunner,
+                0, renderRunner.isLive ? 100 : 1, TimeUnit.MILLISECONDS)};
 
-        Render render = new Render(data, frame);
 
         JPanel regionsPanel = new JPanel() {
             public void paint(Graphics g) {
-                render.renderRegions(g);
+                renderRunner.renderRegions(g);
             }
         };
 
         JPanel legendPanel = new JPanel() {
             @Override
             public void paint(Graphics g) {
-                Render.renderLegend(g);
+                renderRunner.renderLegend(g);
             }
         };
 
         JPanel statusPanel = new JPanel() {
             @Override
             public void paint(Graphics g) {
-                render.renderStats(g);
+                renderRunner.renderStats(g);
             }
         };
 
         JPanel graphPanel = new JPanel() {
             @Override
             public void paint(Graphics g) {
-                render.renderGraph(g);
```

Comment:
I know that this is pre-existing (afaict), but I would strongly recommend to not render the complex stuff directly on the frame. In my experience, the animation goes much smoother (less flickering, better performance, less blocking of event dispatch thread), if the scene is rendered on an offscreen image (using a separate thread), and then, when redrawing is requested on the frame (possibly requested by the render thread, or whenever the GUI thinks it needs to redraw), only draw the offscreen image on the frame. (This could be further improved by not actually drawing on the offscreen image on the frame, but instead flip the buffers, see here for more details on this: https://docs.oracle.com/javase/tutorial/extra/fullscreen/doublebuf.html .) Maybe this should be done as a follow-up change, and not here, though,

Reply:
I will look into this. Agree it should be a separate change.